src/IssueViz/Editor/LocationTagging/LocationTagger.cs

```diff
@@ -0,0 +1,188 @@
+        }
+
+        #region ITagger implementation
+
+        public event EventHandler<SnapshotSpanEventArgs> TagsChanged;
+
+        public IEnumerable<ITagSpan<IIssueLocationTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+        {
+            if (spans.Count == 0 || TagSpans.Count == 0) { yield break; }
+
+            // If the requested snapshot isn't the same as our tags are on, translate our spans to the expected snapshot 
+            if (spans[0].Snapshot != TagSpans[0].Span.Snapshot)
+            {
+                TranslateTagSpans(spans[0].Snapshot);
+            }
+
+            // Find any tags in that overlap with that range
+            foreach (var tagSpan in TagSpans)
+            {
+                if (OverlapsExists(tagSpan.Span, spans))
+                {
+                    yield return tagSpan;
+                }
+            }
+        }
+
+        private void TranslateTagSpans(ITextSnapshot newSnapshot)
+        {
+            var translatedTagSpans = new List<ITagSpan<IIssueLocationTag>>();
+
+            foreach (var old in TagSpans)
+            {
+                var newSpan = old.Span.TranslateTo(newSnapshot, SpanTrackingMode.EdgeExclusive);
+                old.Tag.Location.Span = newSpan; // update the span stored in the location visualization 
+
+                translatedTagSpans.Add(new TagSpan<IIssueLocationTag>(newSpan, old.Tag));
+            }
+
+            TagSpans = translatedTagSpans;
+        }
+
+        private static bool OverlapsExists(Span span, NormalizedSnapshotSpanCollection spans) =>
+            spans.Any(x => x.OverlapsWith(span));
+
+        #endregion ITagger implementation
+
+        #region Buffer change tracking
+
+        internal /* for testing */ string FilePath { get; }
+
+        private void SafeOnBufferChange(object sender, TextContentChangedEventArgs e)
+        {
+            // Handles callback from VS. Suppress non-critical errors to prevent them
+            // propagating to VS, which would display a dialogue and disable the extension.
+            try
+            {
+                // The text buffer has been edited (i.e.text added, deleted or modified).
+                // The spans we have stored for location relate to the previous text buffer and
+                // are no longer valid, so we need to translate them to the equivalent spans
+                // in the new text buffer.
+                var newTextSnapshot = e.After;
+
+                TranslateTagSpans(newTextSnapshot);
+                locationService.Refresh(new string[] { FilePath });
+
+                var wholeSpan = new SnapshotSpan(newTextSnapshot, 0, newTextSnapshot.Length);
+                TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(wholeSpan));
+            }
+            catch (Exception ex) when (!ErrorHandler.IsCriticalException(ex))
+            {
+                logger.WriteLine(Resources.ERR_HandlingBufferChange, ex.Message);
```

Comment:
If there was an exception, what does this mean regarding the existing `TagSpans` property, and the spans stored in each locationViz? Is it worth clearing those?

Reply:
It's hard to say - depends what the exception was and where it occurred.
This part of the implementation is the same as the old one from the `TextBufferIssueTracker` and is purely defensive to stop us crashing VS. We didn't do anything with the existing spans in this case previously, and we didn't have any reports of tags becoming corrupted.